### ISSUE TITLE
Support for adding data watchpoints 

### ISSUE THREAD

**Issue by RaamyPi** [riscv/riscv-openocd#1151]
According to the latest debug spec (dated May 15, 2024), the select field in mcontrol and mcontrol6 provides an option to configure a watchpoint to break on address/data but openocd doesn't have any reference to this field.

**Comment by TommyMurphyTM1234** [comment]
Doesn't `riscv.c` reference/use these?

- https://github.com/riscv-collab/riscv-openocd/blob/riscv/src/target/riscv/riscv.c

**Comment by RaamyPi** [comment]
Here is the function in discussion: [https://github.com/riscv-collab/riscv-openocd/blob/riscv/src/target/riscv/riscv.c#L1514](url)

As you can observe, there is no reference to the 'select' field.

**Comment by en-sc** [comment]
@RaamyPi, AFAIK neither GDB, nor OpenOCD provides means to set a watchpoint on data values (i.e. there is no "select" in OpenOCD's `wp` or GDB's `watch/awatch`.

If you wish for this feature to be supported (on the level of `wp` command) consider posting this question to mainline OpenOCD: https://openocd.org/pages/mailing-lists.html

In the meantime nothing prevents you from setting the watchpoint on data values manually, by reserving the trigger (`riscv reserve_trigger` command) and programming the trigger with accesses to `tselect`, `tdata1`, `tdata2` etc.

**Comment by TommyMurphyTM1234** [comment]
> @RaamyPi, AFAIK neither GDB, nor OpenOCD provides means to set a watchpoint on data values (i.e. there is no "select" in OpenOCD's `wp` or GDB's `watch/awatch`.

Hi @en-sc, I think you're correct about GDB:

- https://sourceware.org/gdb/current/onlinedocs/gdb.html/Set-Watchpoints.html

but doesn't OpenOCD support this via the `value` argument? 

-  https://openocd.org/doc/html/General-Commands.html

> Command: `wp [address length [(r|w|a) [value [mask]]]]`
>
> With no parameters, lists all active watchpoints. Else sets a data watchpoint on data from `address` for `length` bytes. The watch point is an "access" watchpoint unless the `r` or `w` parameter is provided, defining it as respectively a read or write watchpoint. If a `value` is provided, that value is used when determining if the watchpoint should trigger. The value may be first be masked using `mask` to mark “don’t care” fields.

**Comment by en-sc** [comment]
@TommyMurphyTM1234, thanks! I've missed it.

Then it's a bit easier. RISC-V targets can provide support for OpenOCD watchpoints with `value` specified by chaining a watchpoint with `select = address` and a watchpoint with `select = data`.

**Comment by TommyMurphyTM1234** [comment]
FWIW this code seems to explicitly check for and reject watchpoints that specify a data value to match:

- https://github.com/riscv-collab/riscv-openocd/blob/a4020f1a0207dc218812ee262ce861d81aa7962f/src/target/riscv/riscv.c#L1528-L1533

> Then it's a bit easier. RISC-V targets can provide support for OpenOCD watchpoints with value specified by chaining a watchpoint with select = address and a watchpoint with select = data.

It's certainly feasible but I'm not sure about "easy"? At the moment the code seems to be predicated on using one trigger per watchpoint (and hardware breakpoint?) whereas  this change would require additional trigger resource management and availability checking in order to deal with the situation where two triggers are required for a data watchpoint with a value match?

**Comment by en-sc** [comment]
> At the moment the code seems to be predicated on using one trigger per watchpoint (and hardware breakpoint?)...

Not exactly. RISC-V OpenOCD already uses chained triggers, e.g.  to set an unaligned watchpoint:
https://github.com/riscv-collab/riscv-openocd/blob/a4020f1a0207dc218812ee262ce861d81aa7962f/src/target/riscv/riscv.c#L1067

**Comment by TommyMurphyTM1234** [comment]
> > At the moment the code seems to be predicated on using one trigger per watchpoint (and hardware breakpoint?)...
> 
> Not exactly. RISC-V OpenOCD already uses chained triggers, e.g. to set an unaligned watchpoint:
> 
> https://github.com/riscv-collab/riscv-openocd/blob/a4020f1a0207dc218812ee262ce861d81aa7962f/src/target/riscv/riscv.c#L1067

Ah - OK - thanks. I wasn't aware of that. 👍

**Comment by RaamyPi** [comment]
> If you wish for this feature to be supported (on the level of wp command) consider posting this question to mainline OpenOCD: https://openocd.org/pages/mailing-lists.html

@en-sc thank you, I shall follow this up.

> In the meantime nothing prevents you from setting the watchpoint on data values manually, by reserving the trigger (riscv reserve_trigger command) and programming the trigger with accesses to tselect, tdata1, tdata2 etc.

@en-sc you're right and I have already tried this as well but this is impractical as gdb is not aware of these artifical watchpoints.

**Comment by TommyMurphyTM1234** [comment]
> @en-sc you're right and I have already tried this as well but this is impractical as gdb is not aware of these artifical watchpoints.

Even if/when OpenOCD is extended to support value option on data watchpoints this may still be the case given that GDB doesn't have such an option. Unless it will have **some** awareness of watchpoints set using something like `monitor wp 0x80004000 w 0xdeadbeef`? But I don't think that it will? If you want full control and awareness of the value option in GDB then GDB itself would probably also need to be extended?

**Comment by RaamyPi** [comment]
> Even if/when OpenOCD is extended to support value option on data watchpoints this may still be the case given that GDB doesn't have such an option. Unless it will have **some** awareness of watchpoints set using something like `monitor wp 0x80004000 w 0xdeadbeef`? But I don't think that it will? If you want full control and awareness of the value option in GDB then GDB itself would probably also need to be extended?

GDB definitely has to be extended as well,  I wanted to take up the openocd side of things before that.

**Comment by fanghuaqi** [comment]
> If you want full control and awareness of the value option in GDB then GDB itself would probably also need to be extended?

Hello @TommyMurphyTM1234 @RaamyPi , gdb can watch on address with expression checked, like below, I just did a example as below in x86 like this `watch *0x555555558014 if *0x555555558014=0x1234`

```shell
(gdb) p &shared_value
$2 = (int *) 0x555555558014 <shared_value>
(gdb) p /x &shared_value
$3 = 0x555555558014
# watch on a address 0x555555558014  and only the address data matched with 0x1234
(gdb) watch *0x555555558014 if *0x555555558014=0x1234
Hardware watchpoint 2: *0x555555558014
(gdb) x /x 0x555555558014
0x555555558014 <shared_value>:  0x00000000
(gdb) c
Continuing.
[New Thread 0x7ffff7d9b700 (LWP 2357669)]
[New Thread 0x7ffff759a700 (LWP 2357670)]
[Switching to Thread 0x7ffff7d9b700 (LWP 2357669)]

Thread 2 "multi_thread_ex" hit Hardware watchpoint 2: *0x555555558014

Old value = 0
New value = <unreadable>
thread_func (arg=0x7fffffffda40) at test.c:8
8           for (int i = 0; i < 10; i++) {
(gdb) x /x 0x555555558014
0x555555558014 <shared_value>:  0x00001234
```

Or the gdb backend need to support such feature?

**Comment by TommyMurphyTM1234** [comment]
> Hello [@TommyMurphyTM1234](https://github.com/TommyMurphyTM1234) [@RaamyPi](https://github.com/RaamyPi) , gdb can watch on address with expression checked, like below, I just did a example as below in x86 like this `watch *0x555555558014 if *0x555555558014=0x1234`

Yes - but the mechanism used to implement that is most likely repeated reading of the target memory to check if the watch conditions have been met and while this may be pragmatic on a native host platform it is most likely not so in a remote embedded target situation where communication is over a limited bandwidth/speed connection such as JTAG. In the latter case such a mechanism almost certainly needs target side support (e.g. via the RISC-V Debug Module) as well as customisation of, say, OpenOCD (and maybe even GDB?), to implement something that is practical to use.

**Comment by fanghuaqi** [comment]
Yes, @TommyMurphyTM1234  the gdb `Z-packet` as described in https://sourceware.org/gdb/current/onlinedocs/gdb.html/Packets.html#Packets seems not to support watchpoint with condition, do you know whether there is a feature request in gdb to support watchpoint with condition?

**Comment by TommyMurphyTM1234** [comment]
> Yes, [@TommyMurphyTM1234](https://github.com/TommyMurphyTM1234) the gdb `Z-packet` as described in https://sourceware.org/gdb/current/onlinedocs/gdb.html/Packets.html#Packets seems not to support watchpoint with condition, do you know whether there is a feature request in gdb to support watchpoint with condition?

No - you'd need to check the GDB mailing lists/bugbase/community.
